File: src/webengine/libqwebenginehttprequest.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qwebenginehttprequest_enums = enums;
const std = @import("std");
pub const map_constu8_constu8 = std.StringHashMapUnmanaged([]const u8);

/// https://doc.qt.io/qt-6/qwebenginehttprequest.html
pub const qwebenginehttprequest = struct {
    /// New constructs a new QWebEngineHttpRequest object.
    ///
    ///
    pub fn New() QtC.QWebEngineHttpRequest {
        return qtc.QWebEngineHttpRequest_new();
    }

    /// New2 constructs a new QWebEngineHttpRequest object.
    ///
    /// ``` other: QtC.QWebEngineHttpRequest ```
    pub fn New2(other: ?*anyopaque) QtC.QWebEngineHttpRequest {
        return qtc.QWebEngineHttpRequest_new2(@ptrCast(other));
    }

    /// New3 constructs a new QWebEngineHttpRequest object.
    ///
    /// ``` url: QtC.QUrl ```
    pub fn New3(url: ?*anyopaque) QtC.QWebEngineHttpRequest {
        return qtc.QWebEngineHttpRequest_new3(@ptrCast(url));
    }

    /// New4 constructs a new QWebEngineHttpRequest object.
    ///
    /// ``` url: QtC.QUrl, method: *const qwebenginehttprequest_enums.Method ```
    pub fn New4(url: ?*anyopaque, method: *const i32) QtC.QWebEngineHttpRequest {
        return qtc.QWebEngineHttpRequest_new4(@ptrCast(url), @ptrCast(method));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#operator-eq)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest, other: QtC.QWebEngineHttpRequest ```
    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QWebEngineHttpRequest_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#postRequest)
    ///
    /// ``` url: QtC.QUrl, postData: map_constu8_constu8, allocator: std.mem.Allocator ```
    pub fn PostRequest(url: ?*anyopaque, postData: map_constu8_constu8, allocator: std.mem.Allocator) QtC.QWebEngineHttpRequest {
        const postData_keys = allocator.alloc(qtc.libqt_string, postData.count()) catch @panic("qwebenginehttprequest.PostRequest: Memory allocation failed");
        defer allocator.free(postData_keys);
        const postData_values = allocator.alloc([]const u8, postData.count()) catch @panic("qwebenginehttprequest.PostRequest: Memory allocation failed");
        defer allocator.free(postData_values);
        var i: usize = 0;
        var postData_it = postData.iterator();
        while (postData_it.next()) |entry| {
            const key = entry.key_ptr.*;
            postData_keys[i] = qtc.libqt_string{
                .len = key.len,
                .data = key.ptr,
            };
            postData_values[i] = entry.value_ptr.*;
            i += 1;
        }
        const postData_map = qtc.libqt_map{
            .len = postData.count(),
            .keys = @ptrCast(postData_keys.ptr),
            .values = @ptrCast(postData_values.ptr),
        };
        return qtc.QWebEngineHttpRequest_PostRequest(@ptrCast(url), postData_map);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#swap)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest, other: QtC.QWebEngineHttpRequest ```
    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QWebEngineHttpRequest_Swap(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#operator-eq-eq)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest, other: QtC.QWebEngineHttpRequest ```
    pub fn OperatorEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QWebEngineHttpRequest_OperatorEqual(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#operator-not-eq)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest, other: QtC.QWebEngineHttpRequest ```
    pub fn OperatorNotEqual(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QWebEngineHttpRequest_OperatorNotEqual(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#method)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest ```
    ///
    /// Returns: ``` qwebenginehttprequest_enums.Method ```
    pub fn Method(self: ?*anyopaque) i32 {
        return qtc.QWebEngineHttpRequest_Method(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#setMethod)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest, method: qwebenginehttprequest_enums.Method ```
    pub fn SetMethod(self: ?*anyopaque, method: i32) void {
        qtc.QWebEngineHttpRequest_SetMethod(@ptrCast(self), @intCast(method));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#url)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest ```
    pub fn Url(self: ?*anyopaque) QtC.QUrl {
        return qtc.QWebEngineHttpRequest_Url(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#setUrl)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest, url: QtC.QUrl ```
    pub fn SetUrl(self: ?*anyopaque, url: ?*anyopaque) void {
        qtc.QWebEngineHttpRequest_SetUrl(@ptrCast(self), @ptrCast(url));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#postData)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest, allocator: std.mem.Allocator ```
    pub fn PostData(self: ?*anyopaque, allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QWebEngineHttpRequest_PostData(@ptrCast(self));
        defer qtc.libqt_string_free(&_bytearray);
        const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qwebenginehttprequest.PostData: Memory allocation failed");
        @memcpy(_ret, _bytearray.data[0.._bytearray.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#setPostData)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest, postData: []u8 ```
    pub fn SetPostData(self: ?*anyopaque, postData: []u8) void {
        const postData_str = qtc.libqt_string{
            .len = postData.len,
            .data = postData.ptr,
        };
        qtc.QWebEngineHttpRequest_SetPostData(@ptrCast(self), postData_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#hasHeader)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest, headerName: []u8 ```
    pub fn HasHeader(self: ?*anyopaque, headerName: []u8) bool {
        const headerName_str = qtc.libqt_string{
            .len = headerName.len,
            .data = headerName.ptr,
        };
        return qtc.QWebEngineHttpRequest_HasHeader(@ptrCast(self), headerName_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#headers)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest, allocator: std.mem.Allocator ```
    pub fn Headers(self: ?*anyopaque, allocator: std.mem.Allocator) [][]u8 {
        const _arr: qtc.libqt_list = qtc.QWebEngineHttpRequest_Headers(@ptrCast(self));
        const _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i| {
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            }
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]u8, _arr.len) catch @panic("qwebenginehttprequest.Headers: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data = _str[i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("qwebenginehttprequest.Headers: Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#header)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest, headerName: []u8, allocator: std.mem.Allocator ```
    pub fn Header(self: ?*anyopaque, headerName: []u8, allocator: std.mem.Allocator) []u8 {
        const headerName_str = qtc.libqt_string{
            .len = headerName.len,
            .data = headerName.ptr,
        };
        const _bytearray: qtc.libqt_string = qtc.QWebEngineHttpRequest_Header(@ptrCast(self), headerName_str);
        defer qtc.libqt_string_free(&_bytearray);
        const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qwebenginehttprequest.Header: Memory allocation failed");
        @memcpy(_ret, _bytearray.data[0.._bytearray.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#setHeader)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest, headerName: []u8, value: []u8 ```
    pub fn SetHeader(self: ?*anyopaque, headerName: []u8, value: []u8) void {
        const headerName_str = qtc.libqt_string{
            .len = headerName.len,
            .data = headerName.ptr,
        };
        const value_str = qtc.libqt_string{
            .len = value.len,
            .data = value.ptr,
        };
        qtc.QWebEngineHttpRequest_SetHeader(@ptrCast(self), headerName_str, value_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#unsetHeader)
    ///
    /// ``` self: QtC.QWebEngineHttpRequest, headerName: []u8 ```
    pub fn UnsetHeader(self: ?*anyopaque, headerName: []u8) void {
        const headerName_str = qtc.libqt_string{
            .len = headerName.len,
            .data = headerName.ptr,
        };
        qtc.QWebEngineHttpRequest_UnsetHeader(@ptrCast(self), headerName_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qwebenginehttprequest.html#dtor.QWebEngineHttpRequest)
    ///
    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.QWebEngineHttpRequest ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QWebEngineHttpRequest_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qwebenginehttprequest.html#types
pub const enums = struct {
    pub const Method = enum {
        pub const Get: i32 = 0;
        pub const Post: i32 = 1;
    };
};

File: src_original/webengine/libqwebenginehttprequest.cpp
#include <QByteArray>
#include <QList>
#include <QMap>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QUrl>
#include <QWebEngineHttpRequest>
#include <qwebenginehttprequest.h>
#include "libqwebenginehttprequest.h"
#include "libqwebenginehttprequest.hxx"

QWebEngineHttpRequest* QWebEngineHttpRequest_new() {
    return new QWebEngineHttpRequest();
}

QWebEngineHttpRequest* QWebEngineHttpRequest_new2(const QWebEngineHttpRequest* other) {
    return new QWebEngineHttpRequest(*other);
}

QWebEngineHttpRequest* QWebEngineHttpRequest_new3(const QUrl* url) {
    return new QWebEngineHttpRequest(*url);
}

QWebEngineHttpRequest* QWebEngineHttpRequest_new4(const QUrl* url, const int* method) {
    return new QWebEngineHttpRequest(*url, (const QWebEngineHttpRequest::Method&)(*method));
}

void QWebEngineHttpRequest_OperatorAssign(QWebEngineHttpRequest* self, const QWebEngineHttpRequest* other) {
    self->operator=(*other);
}

QWebEngineHttpRequest* QWebEngineHttpRequest_PostRequest(const QUrl* url, const libqt_map /* of libqt_string to libqt_string */ postData) {
    QMap<QString, QString> postData_QMap;
    libqt_string* postData_karr = static_cast<libqt_string*>(postData.keys);
    libqt_string* postData_varr = static_cast<libqt_string*>(postData.values);
    for (size_t i = 0; i < postData.len; ++i) {
        QString postData_karr_i_QString = QString::fromUtf8(postData_karr[i].data, postData_karr[i].len);
        QString postData_varr_i_QString = QString::fromUtf8(postData_varr[i].data, postData_varr[i].len);
        postData_QMap[postData_karr_i_QString] = postData_varr_i_QString;
    }
    return new QWebEngineHttpRequest(QWebEngineHttpRequest::postRequest(*url, postData_QMap));
}

void QWebEngineHttpRequest_Swap(QWebEngineHttpRequest* self, QWebEngineHttpRequest* other) {
    self->swap(*other);
}

bool QWebEngineHttpRequest_OperatorEqual(const QWebEngineHttpRequest* self, const QWebEngineHttpRequest* other) {
    return (*self == *other);
}

bool QWebEngineHttpRequest_OperatorNotEqual(const QWebEngineHttpRequest* self, const QWebEngineHttpRequest* other) {
    return (*self != *other);
}

int QWebEngineHttpRequest_Method(const QWebEngineHttpRequest* self) {
    return static_cast<int>(self->method());
}

void QWebEngineHttpRequest_SetMethod(QWebEngineHttpRequest* self, int method) {
    self->setMethod(static_cast<QWebEngineHttpRequest::Method>(method));
}

QUrl* QWebEngineHttpRequest_Url(const QWebEngineHttpRequest* self) {
    return new QUrl(self->url());
}

void QWebEngineHttpRequest_SetUrl(QWebEngineHttpRequest* self, const QUrl* url) {
    self->setUrl(*url);
}

libqt_string QWebEngineHttpRequest_PostData(const QWebEngineHttpRequest* self) {
    QByteArray _qb = self->postData();
    libqt_string _str;
    _str.len = _qb.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _qb.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QWebEngineHttpRequest_SetPostData(QWebEngineHttpRequest* self, const libqt_string postData) {
    QByteArray postData_QByteArray(postData.data, postData.len);
    self->setPostData(postData_QByteArray);
}

bool QWebEngineHttpRequest_HasHeader(const QWebEngineHttpRequest* self, const libqt_string headerName) {
    QByteArray headerName_QByteArray(headerName.data, headerName.len);
    return self->hasHeader(headerName_QByteArray);
}

libqt_list /* of libqt_string */ QWebEngineHttpRequest_Headers(const QWebEngineHttpRequest* self) {
    QList<QByteArray> _ret = self->headers();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QByteArray _lv_qb = _ret[i];
        libqt_string _lv_str;
        _lv_str.len = _lv_qb.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_qb.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_string QWebEngineHttpRequest_Header(const QWebEngineHttpRequest* self, const libqt_string headerName) {
    QByteArray headerName_QByteArray(headerName.data, headerName.len);
    QByteArray _qb = self->header(headerName_QByteArray);
    libqt_string _str;
    _str.len = _qb.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _qb.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QWebEngineHttpRequest_SetHeader(QWebEngineHttpRequest* self, const libqt_string headerName, const libqt_string value) {
    QByteArray headerName_QByteArray(headerName.data, headerName.len);
    QByteArray value_QByteArray(value.data, value.len);
    self->setHeader(headerName_QByteArray, value_QByteArray);
}

void QWebEngineHttpRequest_UnsetHeader(QWebEngineHttpRequest* self, const libqt_string headerName) {
    QByteArray headerName_QByteArray(headerName.data, headerName.len);
    self->unsetHeader(headerName_QByteArray);
}

void QWebEngineHttpRequest_Delete(QWebEngineHttpRequest* self) {
    delete self;
}
